pg_search/src/index/reader/index.rs

```diff
@@ -209,9 +209,12 @@ impl SearchIndexReader {
         //rendering those dead ctids visible. The concurrent scan then returns the wrong results.
         // To prevent this, ambulkdelete acquires an exclusive cleanup lock. Readers must also acquire this lock (shared)
         // to prevent a reader from reading dead ctids right before ambulkdelete finishes.
+        //
+        // It's sufficient, and **required** for parallel scans to operate correctly, for us to hold onto
+        // a pinned but unlocked buffer.
         let cleanup_lock = if needs_cleanup_lock {
             let bman = BufferManager::new(index_relation.oid());
-            Some(bman.get_buffer(CLEANUP_LOCK))
+            Some(bman.get_buffer_pinned(CLEANUP_LOCK))
```

Comment:
This is actual fix.  `get_buffer_pinned()` is a new function and there's other refactoring that's happened here to help keep things tidy going forward.

Reply:
We need to acquire a share lock here. It can immediately be dropped, but if we don't then a scan could start while the VM is being updated.

Reply:
Ok so this is just in case the visibility map has been updated from a bulk delete and we're waiting for `amvacuumcleanup` to release the lock. Makes sense.

I'm not sure doing the back-to-back LockBuffer calls right here would be ideal? That's the obvious thing to do. But `PinnedBuffer` may be useful in the future, so perhaps a routine to upgrade or downgrade `PinnedBuffer` <-> `Buffer/BufferMut` by acquiring or releasing the share/exclusive locks respectively? Or at least the Buffer/Mut -> PinnedBuffer direction. Then we could do something like this:

```suggestion
            let buffer = bman.get_buffer(CLEANUP_LOCK);
            Some(buffer.unlock()) // PinnedBuffer
```

That would feel clean to me 🤷‍♂️ And maybe call it `UnlockedBuffer`. Of course I may be overthinking this 😆